content/posts/numpy/numpy2/index.md

```diff
@@ -0,0 +1,112 @@
+---
+title: "18 years in the Making: NumPy 2.0 Released"
+date: 2024-06-16
+draft: false
+description: "
+18 years in the Making: NumPy 2.0 Released
+"
+tags: ["News", "numpy"]
+displayInList: true
+author: ["NumPy Developers"]
+---
+
+After 18 years since the release of NumPy 1.0, we are thrilled to announce the
+launch of NumPy 2.0! This major release marks a significant milestone in the
+evolution of NumPy, bringing a wealth of enhancements and improvements to users
+and unlocking future developments.
+
+NumPy has improved and evolved over the past 18 years and many old releases had
+significant performance, usability, and consistency improvements.
+The journey to an actual 2.0 release has been long, and it was difficult to
+build the necessary momentum. In part this may be because for a time the idea
+of doing a NumPy 2.0 release among the NumPy developers required a revolutionary change with
+rewrites of significant key pieces in the code base. Many of these rewrites and
```

Comment:
Maybe something like this following "... past 18 years".

"our  approach to has been to make incremental changes while keeping backward compatibility. The downside of that approach was an accumulation of outdated functions and duplicated functionality. This approach came to an end with the need to make the type system more extendable. The type system is fundamental to NumPy and major changes could not be made incrementally, mixing two different type systems would be a recipe for disaster.  Once the decision to jump the version was made, it was seen as an opportunity to deal with the accumulated technical debt, discard duplicate functionality, improve the API to enable future developments, and make a clean distinction between the public interface and the private implementation details."

Reply:
I like this. It's quite detailed, but I think it's insightful for the reader who has a "was this really really necessary" question in mind. 

I'm not 100% sure it is fully accurate though, since the making the C API for new dtypes was a very late addition to 2.0; there may have been at least an equal amount of "this is getting more and more painful, let's do a big cleanup just once" energy. Even if that's the case, the "two type systems" rationale is cleaner and easy to understand - so I still like it.

@seberg WDYT about this proposed text?

Reply:
TLDR:  I like the idea also, but happy either way and maybe it needs a slight rephrase because I think both Ralf and I read it maybe not as intended.

---

I had similar thoughts and was thinking about it.  I like the start a lot, but feel talking too much about the type system is only half right.

> since the making the C API for new dtypes was a very late addition 

Making it public has maybe not that much to do with 2.0 itself really.  The only direct relation is that it seems like the right time and `StringDType` helped a bit boost confidence (and probably more...).

OTOH, it is true that e.g. NEP 50 and (even tiny) C-API removals mean that a lot of duplication can now be removed again on the C implementation side (or at least eventually, and some of that is 10+ years old).

> The type system is fundamental to NumPy and major changes could not be made incrementally, mixing two different type systems would be a recipe for disaster

But, @rgommers, I think maybe we misread this and it should refer to NEP 50 user facing behavior and not to the implementation :).
It is not incorrect that NEP 50 was maybe the piece that helped get things moving, since it is so obviously not possible to do this without a major change.

Maybe that can be rephrased (or slightly reordered) to make it more obvious that the behavior change is just as important (or more)?  OTOH, I wonder if then start duplicating the bullet points later.

Reply:
Yeah, the history is sort of made up as I'm not that familiar with it. Modify as needed.

Reply:
Adding those two sentences (or replacing some others with it) is fine with me.  For me, this is really more about behavior than the code (and it reads a bit like code).  But that may be me, and the two go at least in part hand-in-hand.

Reply:
I gave it a go - please have a look at the new commit

Reply:
good point, tried to tweak the phrasing and remove parentheses. I wanted to make clear that we didn't change our BC policy permanently

Reply:
Can you resolve the comment @seberg or @charris? I don't have permissions to do so.